File: 5.2-integration/1.2-function-reconstruction/py/main.py
import math
from input import input_float

# Task 2, 3, 4
def evaluate(x, c):
    print("Reconstructed Function:")
    if(c >= 0):print("3x^2 + 4x +", c)
    else: print("3x^2 + 4x -", c*-1)
    eqn = 3 * x ** 2 + 4 * x + c
    print("Result:", eqn)
    return eqn

def find_c(f_x, x):
    c = f_x - (3 * x ** 2 + 4 * x)
    print("C is:", c)

    return c

def verify(x, f_x, c):
    eqn = 3 * x ** 2 + 4 * x + c
    if(math.isclose(f_x, eqn)): print("Initial condition verified.")
    else: print("Initial condition not verified.")

def main():
    x: float = input_float("Enter initial x:")
    f_x: float = input_float("Enter initial f(x):")
    c= find_c(f_x, x)
    new_x = input_float("Enter value of x: ")
    evaluate(new_x, c)
    verify(x, f_x, c)
    

if __name__ == "__main__":
    main()

File: 5.2-integration/1.2-function-reconstruction/cpp/main.cpp
#include <iostream>
#include <cmath>
#include "input.h"

using namespace std;

float evaluate(float x, float c=0){
    float eqn = 3 * x * x + 4 * x + c;
    return eqn;
}

float find_c(float x, float f_x){
    float c = f_x  - evaluate(x);
    return c;
}

bool verify_c(float x, float f_x, float c){
    float eqn = 3 * x * x + 4 * x + c;
    if(fabs(f_x - eqn) < 1e-5) return true;
    else return false;
}

int main(){
    string der_str = "6x + 4\n";
    cout << "Derivative: " << der_str;

    float i_x = input_float("Enter initial x: ");
    float f_x = input_float("Enter initial f(x): ");

    float c = find_c(i_x, f_x);
    
    cout << "Initial condition: \n";
    cout << "f(" << i_x << ") = " << c << "\n";

    float x = input_float("Enter x: ");

    cout << "Reconstructed function: \n";

    if(fabs(c) < 1e-5) cout << "f(x) = 3x^2 + 4x" << "\n";
    else cout << "f(x) = 3x^2 + 4x" << (c < 0 ? " - " : " + ") << abs(c) << "\n";

    cout << "Result: " << evaluate(x, c) << "\n";

    cout << (verify_c(i_x, f_x, c) ? "Initial condition verified." : "Initial not condition verified.") << "\n";
}